use monotonic clock for all time computations

rlefevre opened the pull request:
**netdata** currently uses a realtime clock for all time computations (sleep times, updated values time checks, ...), this leads to several issues when the system date/time changes brutally (date command, ntpdate).

This patchset is a draft to try to resolve this issue.

The usual good practice is too use a monotonic clock for all time computations to avoid this, this is what this patch does except for a few places when a realtime clock is really needed (and not dangerous).

First, let's see the impacts on current master when we change the system date a few years in the future, then going back (I also added a log in sleep_usec to illustrate more issues):

With release 1.2, netdata was taking 100% cpu on one core after the date change, with release 1.4, this does not seem to be the case anymore but there are still some issues, a few examples:

```
2016-12-01 17:18:06: apps.plugin: INFO: sleeping 0.989672000
2016-12-01 17:18:07: apps.plugin: INFO: sleeping 0.991512000
2020-01-01 00:00:00: netdata: Resuming logging from process 'netdata' (prevented 3669 logs in the last 1200 seconds).
2020-01-01 00:00:00: netdata: INFO: sensors.coretemp_isa_0000_temperature: took too long to be updated (97224112.995 secs). Reseting it.
2020-01-01 00:00:00: netdata: INFO: sensors.radeon_pci_0100_temperature: took too long to be updated (97224112.995 secs). Reseting it.
2020-01-01 00:00:00: netdata: INFO: netdata.plugin_pythond_sensors: took too long to be updated (97224112.995 secs). Reseting it.
2020-01-01 00:00:00: netdata: INFO: cpu.cpufreq: took too long to be updated (97224112.995 secs). Reseting it.
...
```

then going back:

```
2016-12-01 17:19:00: apps.plugin: INFO: sleeping 97224112.004799000
```
More generally, it is very difficult to know the impact everywhere in the code and to be robust when using a realtime clock.

By using a monotonic clock (clock_gettime(CLOCK_MONOTONIC, ...)), we don't have anymore issues, however the main drawback is that graphs time labels and values are not based on current time anymore (monotonic clock is usually closed to uptime). This could make system issues investigations more difficult.

The approach taken is to transmit to the web client the current monotonic time and to use the graphs formatters to add the offset with the web client time, so all graphs are displayed based on the client time (which could also be usefull for a system without a correct date, for example embedded systems without a CMOS battery).

The main drawbacks I see are:
* on reboot, the clock restarts from 0, so rrd values go in the past.  I have not checked yet but this could raise issues with the cache
* this is implemented only for dygraph for now
* could this raise issues when monitoring several systems ?

Notes:
* we could add a setting in the web client to choose between monotonic or realtime clock time value in the graphs
* I didn't change profiles/*.c files as they don't seem to compile anymore, should they be removed ?
* I still use a realtime clock for the following cases, I may have forgotten ones or let too much, this is only a draft:
  * healths/alarms logs (I'm not sure about my modifications there)
  * logs date/time
  * registry expiration date
  * unit tests
* there are likely shell, python or javascript scripts that have the same issues, I have not checked them

Tell me what you think.

Thanks

ktsaou commented:
well... you have done quite some work here!

I think however, that using the monotonic clock in RRD, will introduce a lot more problems compared to the ones it solves.

Consider for example hibernation. So, it runs and then it stops for 1 hour and then it awakes. The dashboard should show gaps on all charts. Using the monotonic clock, this will not happen. We will not be able to tell where the gap is. This information will not be written to the database.

Let's consider NTP, which adjusts the clock continuously but does not affect the monotonic clock. A slightly bad hardware timer, will destroy netdata. After several days the time shown by netdata will be wrong and we will be able to do nothing to solve it.

If for any reason, netdata ends up with a shift in time at the dashboards, it will be **catastrophic**. Our monitoring will be lying to us. We will start taking wrong decisions about the cause of issues.

It sounds scary... and I am not sure we will gain enough from this change, to justify such a risk...

It might be preferable to use the wall clock like we already do, and consult the monotonic clock only for certain decisions, like:

1. prevent disabling error logging
2. detecting jumps in the past #1303, where we should re-set the RRD database
3. delaying alarms #1296 and #1225, where we should delay the alarms check

rlefevre commented:
Well, actually, CLOCK_MONOTONIC is affected by adjustements done by adjtime() and NTP:

```
       CLOCK_MONOTONIC
              Clock that cannot be set and represents monotonic time since some unspecified starting point.  This clock is not affected by discontinuous jumps in the system time (e.g., if the system admin‐
              istrator manually changes the clock), but is affected by the incremental adjustments performed by adjtime(3) and NTP.
```

And if we want to also detect suspend states, we could use CLOCK_BOOTTIME (since 2.6.39):
```
       CLOCK_BOOTTIME (since Linux 2.6.39; Linux-specific)
              Identical  to  CLOCK_MONOTONIC, except it also includes any time that the system is suspended.  This allows applications to get a suspend-aware monotonic clock without having to deal with the
              complications of CLOCK_REALTIME, which may have discontinuities if the time is changed using settimeofday(2).
```

Then we would have the best of both worlds ?

ktsaou commented:
> Well, actually, CLOCK_MONOTONIC is affected by adjustements done by adjtime() and NTP

nice! I didn't know that.

How do you believe we should handle a change in time.

So, let's assume the time is now 12:00:00 and an administrator says it is 13:00:00.
What do you believe should happen on the RRD database and the dashboard?

rlefevre commented:
If we use CLOCK_BOOTIME, suspends will naturally create gaps in RRD.  For changes in time, they won't be reprensented in a RRD nor in a graph, but I think this is expected way, a gap would make think that data was not collected during it, which is actually wrong.

ktsaou commented:
ok. let me explain that I do not object to the use of `clock_gettime()` with `CLOCK_BOOTTIME` or any other timer. The opposite actually. We should use `clock_gettime()` and probably `CLOCK_BOOTTIME` is a good timer, when it is available.

The only problem I see is the use of a unit for time that is not wallclock time.

This is where we use time:

1. stored in RRD database
2. stored in registry entries
3. stored in alarm entries
4. future: chart annotations
5. send to backend servers (graphite, opentsdb, or future: a netdata central health server)
6. saved in log files
7. in calculations for iterating precisely (data collections and alarms)
8. in calculations for data interpolation

The main problem we have with a different unit of measurement is storage, that is 1 to 5 above.

For 1, 2, 3 and 4, we need to save and load back these databases. This might happen after a reboot, after a day, or whatever. So, we need to somehow know the wallclock time of the saved data.

For 5 and 6, it is not possible to alter the time, once data are saved/sent.

For 7 and 8 we only care about calculating time differences, so anything can be used, as far as all plugins and netdata use the same clock.

rlefevre commented:
Thank you for your detailed answer. I know I'm biased because I work a lot with embedded systems without CMOS battery that handle the system time very inconsistently and differently than data center servers for example. So from my point of view, wall clocks are not always better upon reboots that monotonic ones, which is why I don't trust them. Also I understand that there are a lot of date/time uses in netdata that most likely require different solutions.

Actually, I first tried to also not use a monotonic clock for RRDs (like for logs or registry), but the RRD timestamps are mixed at a lot of places with time calculations and also have issues with time jumps, so this may be the solution, but this will require more work.

I have not yet an answer to all your points. I have to think more about it.

ktsaou commented:
nice! Looking forward for your suggestions...

rlefevre commented:
Ok, let's try. 

RRDSETs would contain 3 last_updated values to allow other parts of netdata the use the appropriate one:
- last_updated_monotonic: CLOCK_MONOTONIC time
- last_updated_boottime: CLOCK_BOOTTIME time
- last_updated_realtime:  CLOCK_REALTIME time

There would be a very slight offset between them as we cannot compute them simultaneously, but I don't think this is really an issue.

1. To reload RRD databases, we would update the RRDSETs clocks as follows:
```
last_updated_monotonic -= (now_realtime - last_updated_realtime) - (now_monotonic - last_updated_monotonic)
last_updated_boottime -= (now_realtime - last_updated_realtime) - (now_boottime - last_updated_boottime)
```

Negative realtime offsets in this computation should be clamped to 0 (time jumps or CMOS issues we cannot reslove) and we must handle negative values in *_dt() functions as monotonic clocks usually start at 0 at boot. Having the wall clock stored allows us to find values at specific wall clock values in post mortem analysis. All interpolations computations would then use last_updated_boottime.

2. For cookies, we would use CLOCK_REALTIME. We don't have much choice but I don't think that there is much computation there. If the system time changes, at worst there may be too much or not enough expirations, which seems acceptable.

3. For alarms, time displayed in notifications would use CLOCK_REALTIME, but notifications time computation would use CLOCK_MONOTONIC to avoid flood after suspend, as well as alerts for too old data collections.

4. This is not yet implemented, so I'm not sure about the design, but with all last clocks stored in RRDSETs, I don't think there is much issue there. This would be basically the same.

5. I think that only backends formatters should use realtime timestamps (and before/after), as we cannot handle some offset on their side.

6. Logs should still use CLOCK_REALTIME, but there are no computation there I think. Logs flooding checks would use CLOCK_MONOTONIC to ignore hibernation.

7. scheduling/calculations for iterating should most likely use CLOCK_MONOTONIC to avoid floods after a suspend (like alarms).

8. As said previously, data interpolation would use CLOCK_BOOTTIME to keep hibernations visible

At last, for web clients, we can choose, but in my opinion, it is better to compute arrays based on last_updated_boottime and format the dates with an offset in the client to avoid issues when time goes back (as done in my patch, except I used CLOCK_MONOTONIC).


Well, this is not quite simple, and I'm probrably wrong/incomplete on some points, but my patch is a first step in this direction. The issue then is to check that correct clocks are used instead of working around time jumps, but I think this can still be easier, for example by using well named functions (ex: sched_monotime_usec, interpol_boottime_usec, cookie_realtime_msec, etc.) and prefixing or suffixing all timestamps with their clock type.

If you agree with this approach, I can try to complete my patch with it if you want, however if you think differently, I would perfectly understand and let you do what you think is best.

Best Regards

ktsaou commented:
hm... I immediately agree with a few... but I am confused/not sure about others...

Would it be better, to do it step by step?

For example, I suggest to follow these steps:

1. Add 5 time functions `now_realtime()`, `now_monitonic()`, `now_boottime()`, `now()` and `now_timespec()`. The first 3 should return `microseconds` (`usec`), `now()` should return epoch (`seconds`) and `now_timespec()` should fill the supplied `timespec`. For all of them we should have the autoconf mechanism fallback to something. This is crucial, for it to be merged.

2. `time(NULL)` should be searched/replaced with `now()`.

3. `time_usec()` should be searched/replaced with `now_realtime()`.

4. `gettimeofday()` should be searched/replaced with a `timespec` like you have done and a call to `now_timespec()`.

   If I recall correctly, there is a place netdata needs a `timeval` for `select()` or something. We will have to adapt the code there.

5. Logs flood protection should use `now_monotonic()`. This will immediately solve the issue there.

6. For all delta calculations (data collection iterations, alarm iterations, etc), use `now_monotonic()` to estimate the time for `sleep_usec()`. This should exclusively be used for the calculating the time for `sleep_usec()` - nothing else. It will not help preventing alarm floods though. Alarm floods after hibernation check the last time the rrd database was updated, so these will not be affected by this change.

   This change will be tricky in a few places that use the same variable for multiple purposes.

   It might also be better, to provide a different set of calls for that purpose. Something that will simplify the code a lot, like `hb = heartbeat_init()` and `hb = heartbeat_next(hb, 1000000usec)`. These 2 functions should handle everything for sleeping exactly at heartbeat boundaries, without exposing anything to the caller. `hb` could be a number or a pointer.

Once we reach that point, we will need to discuss the benefits/weaknesses of changing `rrd_done()`, `rrd_next()` and the likes, to something other than realtime. I cannot decide for that change now. We need to have a clear scope/reason for doing it, as it seems quite risky.

How does that sound?

rlefevre commented:
This sounds perfectly reasonable. 

My intention was not to submit a big and risky patch at once, but more to provide a draft to start the discussion. Now that this is done, it's clear that the only way is to progress patch by patch incrementally to be able to review/discuss them properly.

rlefevre commented:
I'm working on this. I have step 1 implemented and will likely have a few more steps pushed for review before the end of week. I have however a few doubts on step 1:

* Your proposed function names are not consistent. Some have the type of clock (real-time, monotonic, ...) as suffixes, other have the unit (now_timespec). I think this could be confusing. 
* `now` as a function name is complicated because it is a very handy variable name, which is actually used a lot in netdata
* I would like to introduce a usec_t type for functions returning microseconds. I'm not fond of new typedef, but this could ease modifications following discussions on the returned type of these functions, and also improve code readability if we don't include the unit in the function name, do you agree?
* Do you agree to have. `now_monotonic()` fall backing to `now_realtime` and `now_boottime` to `now_monotonic`?

Without better idea, I would propose the following API which is quite verbose but at least clear:
* `now_reatime_sec`
* `now_realtime_usec`
* `now_monotonic_usec`
* `now_boottime_usec`

Do you have a better idea or do you want your original proposition (if so, how do I rename `now` variables)?

ktsaou commented:
I agree with everything you propose.
You are the best!

ktsaou commented:
The only problem with `now_monotonic_usec()` falling back to `now_realtime_usec()` is that we will not be able to detect hibernation... but I guess we can live with it...

rlefevre commented:
Good, you were right to propose to work step by step, this is the best way to do it.
I may not be able to work on this tomorrow, but I will most likely push at least steps 1 and hopefully more on Thursday.

rlefevre commented:
Concerning your worries about fallbacks, unfortunately I think there is no perfect solution. There are some hacks I used in the past using `/proc/uptime` for monotonic or `getrusage`, but they are not really portable and are not very fast. However `clock_gettime()` is now very old (older than `accept4()`?), and most likely only MacOSX ports are concerned. We can still document them properly and improve them later.

rlefevre commented:
`clock_gettime()` is using `struct timespec` that includes a nanoseconds field instead of microseconds for `struct timeval`. I can't decide if it is best to keep all netdata computations in microseconds or to change them to nanoseconds:
* if we keep using microseconds, my first patch that just ports the existing code to the new clock API is minimized, so this is less risky at first, but this adds a division for each call to a time function using `clock_gettime` (except the ones returning seconds or timespec), so performance is slightly affected
* if we now use nanoseconds everywhere, we have a better resolution, but the range is reduced to a few hundred years (still enough I think), and a lot of code is impacted, making the review more difficult and more likely to introduce regressions

What would you prefer?

ktsaou commented:
hm... I prefer to do it with microseconds. I don't think that this call is made so frequently that will have any measurable impact.

rlefevre commented:
Ok, thank you, this seems more reasonable for now.

rlefevre commented:
I pushed the first patches. No one should break compilation and only the last two ones should have impact: 
* `replace remaining usleep() calls by sleep_usec()`: I think those remaining usleep() calls should use sleep_usec(), you will confirm it
* the last one should fix log floods issue on time change or hibernation

Notes:
* for now, I kept the `struct timeval` as you prefer to still use microseconds, so there is no much point using `struct timespec` in code outside of clocks API
* the clock API should have fallbacks but I'm not able to test them. Do you have a system or docker without `clock_gettime()` support? (I remember they were missing in some old 2.4 kernels)

rlefevre commented on src/backends.c:
```diff
@@ -395,19 +395,19 @@ void *backends_main(void *ptr) {
 
-    unsigned long long step_ut = frequency * 1000000ULL;
-    unsigned long long random_ut = now_realtime_usec() % (step_ut / 2);
-    time_t before = (time_t)((now_realtime_usec() - step_ut) / 10000000ULL);
```
Was it a typo (`10 000 000ULL` => there is 7 0s = `10*USEC_PER_SEC`)?

ktsaou replied:
oops! yes it is.
It seems it affects only the first iteration, this is why I haven't found it during testing...

ktsaou commented:
Nice work! You are awesome!
Is this ready for review/testing/merging?

ktsaou commented:
Everything you have done is excellent.
The only issue I see is the use of `usec_t` in `apps_plugin.c` for these variables:

```c
usec_t minflt_raw;
usec_t cminflt_raw;
usec_t majflt_raw;
... etc ...
```

this is somewhat misleading, since these variables are not time. They are the rate of something, but not time.

I will start testing this on different systems now...

ktsaou commented on src/rrd.c:
```diff
@@ -998,33 +998,31 @@ collected_number rrddim_set(RRDSET *st, const char *id, collected_number value)
         // microseconds has the time since the last collection
-        unsigned long long now_usec = timeval_usec(&now);
-        unsigned long long last_usec = timeval_usec(&st->last_collected_time);
-        unsigned long long since_last_usec = usec_dt(&now, &st->last_collected_time);
```
these variables break compilation:

```
gcc -DHAVE_CONFIG_H -I. -I..  -DVARLIB_DIR="\"/var/lib/netdata\"" -DCACHE_DIR="\"/var/cache/netdata\"" -DCONFIG_DIR="\"/etc/netdata\"" -DLOG_DIR="\"/var/log/netdata\"" -DPLUGINS_DIR="\"/usr/libexec/netdata/plugins.d\"" -DRUN_DIR="\"/var/run/netdata\"" -DWEB_DIR="\"/usr/share/netdata/web\""        -O1 -ggdb -Wall -Wextra -Wformat-signedness -fstack-protector-all -DNETDATA_INTERNAL_CHECKS=1 -pthread -MT rrd.o -MD -MP -MF .deps/rrd.Tpo -c -o rrd.o rrd.c
In file included from common.h:5:0,
                 from rrd.c:1:
rrd.c: In function ‘rrdset_next_usec’:
rrd.c:1032:25: error: ‘last_usec’ undeclared (first use in this function)
             if(unlikely(last_usec + microseconds > now_usec + 1000))
                         ^
../config.h:202:41: note: in definition of macro ‘unlikely’
 #define unlikely(x) __builtin_expect(!!(x), 0)
                                         ^
rrd.c:1032:25: note: each undeclared identifier is reported only once for each function it appears in
             if(unlikely(last_usec + microseconds > now_usec + 1000))
                         ^
../config.h:202:41: note: in definition of macro ‘unlikely’
 #define unlikely(x) __builtin_expect(!!(x), 0)
                                         ^
rrd.c:1032:52: error: ‘now_usec’ undeclared (first use in this function)
             if(unlikely(last_usec + microseconds > now_usec + 1000))
                                                    ^
../config.h:202:41: note: in definition of macro ‘unlikely’
 #define unlikely(x) __builtin_expect(!!(x), 0)
                                         ^
make[2]: *** [Makefile:676: rrd.o] Error 1
make[2]: *** Waiting for unfinished jobs....
mv -f .deps/health.Tpo .deps/health.Po
mv -f .deps/sys_fs_cgroup.Tpo .deps/sys_fs_cgroup.Po
mv -f .deps/registry.Tpo .deps/registry.Po
make[2]: Leaving directory '/home/costa/src/netdata-rlefevre.git/src'
make[1]: *** [Makefile:472: all-recursive] Error 1
make[1]: Leaving directory '/home/costa/src/netdata-rlefevre.git'
make: *** [Makefile:411: all] Error 2
```

ktsaou commented on src/apps_plugin.c:
```diff
@@ -358,28 +358,28 @@ struct pid_stat {
+    usec_t gtime;
+    usec_t cutime;
+    usec_t cstime;
+    usec_t cgtime;
```
These variables measure utilization that is not `usec_t`. They are whatever the system provides (`flt` are page faults, `time` are CPU time in Jiffies or whatever the system uses).

ktsaou commented on src/apps_plugin.c:
```diff
@@ -829,14 +829,14 @@ int read_proc_pid_io(struct pid_stat *p) {
-unsigned long long global_gtime = 0;
+usec_t global_utime = 0;
+usec_t global_stime = 0;
+usec_t global_gtime = 0;
```
These too are not `usec_t`.

ktsaou commented on src/apps_plugin.c:
```diff
@@ -829,14 +829,14 @@ int read_proc_pid_io(struct pid_stat *p) {
     static char filename[FILENAME_MAX + 1] = "";
     static procfile *ff = NULL;
-    static unsigned long long utime_raw = 0, stime_raw = 0, gtime_raw = 0, gntime_raw = 0, ntime_raw = 0, collected_usec = 0, last_collected_usec = 0;
+    static usec_t utime_raw = 0, stime_raw = 0, gtime_raw = 0, gntime_raw = 0, ntime_raw = 0, collected_usec = 0, last_collected_usec = 0;
```
the `_raw` variables too, are not `usec_t`.

ktsaou replied:
the `_raw` variables too, are not `usec_t`.

ktsaou commented on src/apps_plugin.c:
```diff
@@ -848,34 +848,34 @@ int read_proc_stat() {
     file_counter++;
 
-    unsigned long long last;
+    usec_t last;
```
this one too.

ktsaou commented on src/apps_plugin.c:
```diff
@@ -1331,8 +1331,8 @@ void find_lost_child_debug(struct pid_stat *pe, unsigned long long lost, int typ
 
-unsigned long long remove_exited_child_from_parent(unsigned long long *field, unsigned long long *pfield) {
-    unsigned long long absorbed = 0;
+usec_t remove_exited_child_from_parent(usec_t *field, usec_t *pfield) {
```
these too.

ktsaou commented on src/apps_plugin.c:
```diff
@@ -1331,8 +1331,8 @@ void find_lost_child_debug(struct pid_stat *pe, unsigned long long lost, int typ
-unsigned long long remove_exited_child_from_parent(unsigned long long *field, unsigned long long *pfield) {
-    unsigned long long absorbed = 0;
+usec_t remove_exited_child_from_parent(usec_t *field, usec_t *pfield) {
+    usec_t absorbed = 0;
```
and this one.

ktsaou commented on src/apps_plugin.c:
```diff
@@ -1384,7 +1384,7 @@ void process_exited_processes() {
             if(!pp->updated) continue;
 
-            unsigned long long absorbed;
+            usec_t absorbed;
```
and this one.

ktsaou commented:
I think that everything except the last commit: https://github.com/firehol/netdata/pull/1313/commits/fc7001247bfb5bd777adaccf3a7ec4e4342a36e3 is mergeable. So, would you like to remove this commit to merge it and continue on another PR?

rlefevre commented:
I fixed the last commit, hopefully this is ok now and can be merged.

However as said previously, I was not able to test all these patch on a system without `clock_gettime()`, so there is still a risk there with the fallback functions.

ktsaou commented:
hm... testing without `clock_gettime()` is pretty easy if we just make the autoconf detection of it fail. Right?

rlefevre commented:
For glibc >= 2.17, it is included in libc without any other lib needed (before it was usually in librt), so I'm not sure to see an easy way to test without it.

ktsaou commented:
Probably I didn't say it right. What if we comment out these lines in Makefile.am:

```
AC_CHECK_TYPES([struct timespec, clockid_t], [], [], [[#include <time.h>]])
AC_SEARCH_LIBS([clock_gettime], [rt posix4])
AC_CHECK_FUNCS([clock_gettime])
```

ktsaou commented:
oops! sorry. A recent change broke this one.
Could you please resolve the conflict?

rlefevre commented:
I rebased the patches. I also removed the error logs in clock.c after some tests because `error()` function actually uses clocks functions to check for logs flood. This was leading to recursive calls.

At last I did a few tests of the fallback functions and it seems ok, but it is not exactly the same as a system without `timespec`, `clockid_t`, `CLOCK_*` enums and `clock_gettime()` function.

I think this could be merged if you're ok with it.

ktsaou commented:
This seems very good!
You have done an excellent job!

> At last I did a few tests of the fallback functions and it seems ok, but it is not exactly the same as a system without timespec, clockid_t, CLOCK_* enums and clock_gettime() function.

So, you believe that the functions will be there, but netdata will fail to be linked?

We could use m4 macros like I have done for `mallinfo()` or `mallopt()` (https://github.com/firehol/netdata/blob/master/m4/ax_c_mallinfo.m4) to compile and link a very small program using `clock_gettime()`. This will ensure that even linking works.

If you believe the function will be there, it will be linkable, but somehow it will not work, we could provide a `configure` option to disable `clock_gettime()` and use `gettimeofday()`.

ktsaou commented:
I too think this is mergeable. So, once we will settle your second thoughts on this, we will merge it.

rlefevre commented:
No, I think everything should work, because I redefine types if they are not found in system headers, and also `clock_gettime()` using `gettimeofday()`, of course there will be limitations in case of hibernation or system date changes, as you said before, because even monotonic and boot time functions will use the wall clock.

ktsaou commented:
ok. merging it.
